### DISCUSSION TITLE
Getting remote process name over SSH connection

### DISCUSSION THREAD

**Discussion by mrjones2014** [wezterm/wezterm#3648] [Q&A] [answered]
Context: https://github.com/mrjones2014/smart-splits.nvim/issues/103

[pane:get_foreground_process_name()](https://wezfurlong.org/wezterm/config/lua/pane/get_foreground_process_name.html?h=get_foreground_process_name):

> This information is only available for local panes. Multiplexer panes do not report this information. Similarly, if you are using eg: ssh to connect to a remote host, you won't be able to access the name of the remote process that is running.

However, the Wezterm keymaps currently recommended by `smart-splits.nvim` requires knowing the foreground process name to determine whether to consume the key press or pass it on to the nvim process. Is there any way to work around this?

**Comment by wez** [comment, marked as answer]
You can have your nvim integration set an `IS_NVIM` user var when it starts, and clear it when nvim exits.
See: https://wezfurlong.org/wezterm/config/lua/pane/get_user_vars.html

**Reply by mrjones2014** [reply]
Good idea, would that have less latency than getting the process name too? If so that should probably be the default implementation

**Reply by mrjones2014** [reply]
Experimented with this this morning but couldn't get it working -- could it be because commands run via `nvim` run in a subshell? I'm not sure why this wouldn't set the user var.

Would you mind taking a look at this diff? I added `print` statements in my Wezterm config and it always reports that that var is still `nil`: https://github.com/mrjones2014/smart-splits.nvim/compare/mrj/wezterm-user-vars?expand=1

**Reply by wez** [reply]
I'd suggest looking at oscyank to see how they manage to get the OSC sequence working: https://github.com/ojroques/vim-oscyank/blob/main/plugin/oscyank.vim

**Reply by mrjones2014** [reply]
Hmm, I've tried tons of stuff and still couldn't get it working. It looks like that plugin writes it to `stderr`, which I also tried.

I'm not sure what I'm doing wrong, but when I `print` the entire user vars table from my wezterm config, it's always an empty table `{}` 😞

**Reply by mrjones2014** [reply]
I even tried putting it in a bash script and running the bash script via Nvim's job control API.

**Reply by mrjones2014** [reply]
wtf, it works if I write in vimscript but not Lua...

**Reply by mrjones2014** [reply]
Turns out it is because vimscript treats the escapes as octal (correctly) but Lua in Neovim considers them decimal.

**Reply by wez** [reply]
Ah, I recommend using hex rather than octal! Have been bitten by octal in obscure situations several times in the past!

**Reply by relausen** [reply]
This is a _very_ elegant solution, and I have started using it myself. I do have a small problem, though, and I don't know if it's a bug or just how it works :-).

I use the multiplexer using Unix domains, exactly as described in the docs: https://wezfurlong.org/wezterm/multiplexing.html#unix-domains. Using the IS_NVIM user var in my WezTerm Lua code to control if certain keys should be relayed to Neovim generally works like a charm. But if I have Neovim running when detaching from the multiplexer (or just quit WezTerm), and subsequently connects again, the IS_NVIM user var is no longer set, meaning that the keystrokes I want to pass through to Neovim are not passed through.

Is this by design, and if so, would you mind explaining why this is?

My guess is that the user vars are stored on the "client side", not the "server side", but I would like to have an authoritative answer :-)

P.S.: It's not a showstopper in any way. I have just created a Neovim custom command to set the IS_NVIM user var and mapped it to a key :-)

P.P.S.: I would love to just use the pre-defined WEZTERM_PROG user var that is set by the shell integrations, but then the problem described above _would_ be a showstopper ;-)

P.P.P.S.: A nice solution to this would be if I could find a Neovim event to react to in an autocommand when I connect, but I can't for the life of me find one that works.

**Reply by wez** [reply]
@relausen sounds like maybe there is a missing sync of user vars on attach; please file an issue for this so that I don't forget about it

**Reply by relausen** [reply]
Thank you for the fast response! As requested, I have filed an issue. See #5832.

And allow me to say that I really enjoy using WezTerm. I live my dev life in the terminal and this is by far the best terminal emulator I have ever tried.
